File: src/cpp/tests/test_http.cpp
#ifdef FCB_WITH_CURL

#include <doctest/doctest.h>

#include <fcb/cityjson.hpp>
#include <fcb/http/curl_range_reader.hpp>
#include <fcb/reader.hpp>

#include <cstdlib>
#include <memory>
#include <set>
#include <string>

using namespace fcb;

/// Set by run_http_tests.cmake to a loopback URL serving delft.fcb.
static std::string fixture_url() {
    const char* u = std::getenv("FCB_TEST_HTTP_URL");
    return u != nullptr ? std::string(u) : std::string();
}

#define SKIP_IF_NO_SERVER(url)                                  \
    if ((url).empty()) {                                        \
        MESSAGE("FCB_TEST_HTTP_URL not set; skipping");         \
        return;                                                 \
    }

TEST_CASE("CurlRangeReader reports total size") {
    const std::string url = fixture_url();
    SKIP_IF_NO_SERVER(url);

    CurlRangeReader r(url);
    CHECK(r.total_size() == 7668160);
}

TEST_CASE("remote reads return the same bytes as the local file") {
    const std::string url = fixture_url();
    SKIP_IF_NO_SERVER(url);

    FileRangeReader local(FCB_TEST_DATA_DIR "/delft.fcb");
    CurlRangeReader remote(url);

    CHECK(remote.total_size() == local.total_size());
    CHECK(remote.read(0, 64) == local.read(0, 64));
    CHECK(remote.read(1000, 256) == local.read(1000, 256));

    // The tail: exercises a range that ends exactly at EOF.
    const std::uint64_t n = local.total_size();
    CHECK(remote.read(n - 16, 16) == local.read(n - 16, 16));
}

TEST_CASE("a zero-length read never contacts the server") {
    const std::string url = fixture_url();
    SKIP_IF_NO_SERVER(url);

    CurlRangeReader r(url);
    r.total_size();
    r.reset_request_count();
    CHECK(r.read(100, 0).empty());
    CHECK(r.request_count() == 0);
}

TEST_CASE("reading past the end returns empty rather than throwing") {
    const std::string url = fixture_url();
    SKIP_IF_NO_SERVER(url);

    CurlRangeReader r(url);
    const std::uint64_t n = r.total_size();
    CHECK(r.read(n + 1000, 16).empty());
}

TEST_CASE("a range crossing EOF returns exactly the bytes that exist") {
    const std::string url = fixture_url();
    SKIP_IF_NO_SERVER(url);

    CurlRangeReader r(url);
    const std::uint64_t n = r.total_size();
    CHECK(r.read(n - 10, 100).size() == 10);
}

TEST_CASE("a server ignoring Range still yields the correct slice") {
    // The dangerous case: answering 200 with the whole body. Truncating to
    // `length` would return bytes [0,length) instead of [offset,offset+length).
    const std::string url = fixture_url();
    SKIP_IF_NO_SERVER(url);

    FileRangeReader local(FCB_TEST_DATA_DIR "/delft.fcb");
    CurlRangeReader ignoring(url + "?ignore_range=1");

    CHECK(ignoring.read(1000, 32) == local.read(1000, 32));
    CHECK(ignoring.read(0, 8) == local.read(0, 8));
}

TEST_CASE("a malformed Content-Range is rejected") {
    const std::string url = fixture_url();
    SKIP_IF_NO_SERVER(url);

    CurlRangeReader bad(url + "?bad_range=1");
    CHECK_THROWS_AS(bad.read(100, 16), Error);
}

TEST_CASE("a server answering a different range than requested is rejected") {
    // Returning the wrong offset silently would corrupt every downstream
    // read, so the client must verify Content-Range's start.
    const std::string url = fixture_url();
    SKIP_IF_NO_SERVER(url);

    CurlRangeReader wrong(url + "?wrong_offset=1");
    CHECK_THROWS_AS(wrong.read(100, 64), Error);
}

TEST_CASE("opening a remote file over HTTP parses the same header") {
    const std::string url = fixture_url();
    SKIP_IF_NO_SERVER(url);

    FcbReader local = FcbReader::open_file(FCB_TEST_DATA_DIR "/delft.fcb");
    FcbReader remote = FcbReader::open(std::make_shared<CurlRangeReader>(url));

    CHECK(remote.header().info().features_count == local.header().info().features_count);
    CHECK(remote.header().info().cityjson_version == local.header().info().cityjson_version);
    CHECK(remote.header().layout().feature_begin == local.header().layout().feature_begin);
}

TEST_CASE("opening a remote file costs a bounded number of requests") {
    // The whole point of the 12944-byte prefetch: header plus the top R-tree
    // levels should arrive in essentially one round trip.
    const std::string url = fixture_url();
    SKIP_IF_NO_SERVER(url);

    auto r = std::make_shared<CurlRangeReader>(url);
    r->total_size();
    r->reset_request_count();

    FcbReader reader = FcbReader::open(r);
    CHECK(reader.header().info().features_count > 0);
    CHECK(r->request_count() <= 2);
}

TEST_CASE("a bbox query over HTTP returns the same ids as over the local file") {
    const std::string url = fixture_url();
    SKIP_IF_NO_SERVER(url);

    auto ids_from = [](std::shared_ptr<RangeReader> rr) {
        FcbReader r = FcbReader::open(std::move(rr));
        const auto& info = r.header().info();
        const double mid_x =
            (info.geographical_extent[0] + info.geographical_extent[3]) / 2.0;
        BBox half{info.geographical_extent[0], info.geographical_extent[1], mid_x,
                  info.geographical_extent[4]};
        FeatureIterator it = r.select_bbox(half);
        std::set<std::string> ids;
        while (it.next()) ids.insert(it.current().id());
        return ids;
    };

    auto local = ids_from(std::make_shared<FileRangeReader>(FCB_TEST_DATA_DIR "/delft.fcb"));
    auto remote = ids_from(std::make_shared<CurlRangeReader>(url));
    CHECK_FALSE(local.empty());
    CHECK(local == remote);
}

TEST_CASE("CityJSON emitted over HTTP matches the local file exactly") {
    const std::string url = fixture_url();
    SKIP_IF_NO_SERVER(url);

    FcbReader local = FcbReader::open_file(FCB_TEST_DATA_DIR "/delft.fcb");
    FcbReader remote = FcbReader::open(std::make_shared<CurlRangeReader>(url));

    auto lit = local.select_all();
    auto rit = remote.select_all();

    int checked = 0;
    while (checked < 25 && lit.next() && rit.next()) {
        CHECK(to_cityjson_feature(lit.current(), local.header()) ==
              to_cityjson_feature(rit.current(), remote.header()));
        ++checked;
    }
    CHECK(checked == 25);
}

#endif  // FCB_WITH_CURL

File: src/py/tests/test_http.py
from __future__ import annotations

import os
import select
import subprocess
import sys
import time
from pathlib import Path
from typing import Iterator

import pytest
from flatcitybuf.cityjson import to_cityjson_feature
from flatcitybuf.errors import ErrorCode, FcbError
from flatcitybuf.http_reader import HttpRangeReader
from flatcitybuf.keys import KeyValue
from flatcitybuf.packed_rtree import search_rtree
from flatcitybuf.range_reader import FileRangeReader
from flatcitybuf.reader import FcbReader
from flatcitybuf.stree import AttrCondition, Operator

REPO_ROOT = Path(__file__).resolve().parents[3]
CORPUS = REPO_ROOT / "conformance"
DELFT = REPO_ROOT / "examples" / "data" / "delft.fcb"
RANGE_SERVER = REPO_ROOT / "src" / "cpp" / "tests" / "range_server.py"

SMALL = CORPUS / "small.fcb"

# ---------------------------------------------------- live remote fixture ---
#
# Opt-in integration test against the real, published 3DBAG file
# (~68 GB, EPSG:28992). It is SKIPPED unless FCB_REMOTE_HTTP_URL is set --
# CI never touches the network here, and nobody downloads 68 GB by running
# `pytest`. Enable it with `just test-remote` (which sets the env var to the
# default URL below), or point FCB_REMOTE_HTTP_URL at any current-format file.
#
# The expected values were cross-checked across the Rust, C++, Python and
# TypeScript readers on 2026-07-23; all four agree. If the file is
# regenerated they must be updated in lock-step here and in the other three
# suites (src/rust/fcb_core/tests/http.rs, src/cpp/tests/test_http.cpp,
# src/ts/test/http.test.ts).
REMOTE_URL = os.environ.get("FCB_REMOTE_HTTP_URL", "")
REMOTE_DEFAULT_URL = (
    "https://storage.googleapis.com/flatcitybuf/3dbag_all_index.fcb"
)
REMOTE_FEATURES_COUNT = 10_771_547
# A ~1 km box over central Amsterdam (minx, miny, maxx, maxy), well inside
# the national extent.
REMOTE_BBOX = (120_000.0, 486_000.0, 121_000.0, 487_000.0)
REMOTE_BBOX_COUNT = 2762

# ---------------------------------------------------------- server fixture ---
#
# src/cpp/tests/range_server.py is the C++ suite's own range-capable test
# server (real Range/Content-Range handling, plus ?ignore_range=1,
# ?bad_range=1, ?wrong_offset=1, ?no_etag=1 query-param misbehaviours --
# see its own module docstring). Reused here rather than writing a second
# one. It binds port 0 and prints the chosen port on stdout.


def _wait_for_port(proc: subprocess.Popen[str], timeout: float = 10.0) -> int:
    """Read the port range_server.py prints, without ever blocking past
    `timeout` -- a server that fails to start must not hang the suite."""
    deadline = time.monotonic() + timeout
    assert proc.stdout is not None
    while True:
        remaining = deadline - time.monotonic()
        if remaining <= 0:
            proc.kill()
            pytest.fail("range_server.py did not report a port in time")
        if proc.poll() is not None:
            stderr = proc.stderr.read() if proc.stderr else ""
            pytest.fail(
                "range_server.py exited before reporting a port "
                f"(returncode={proc.returncode}): {stderr}"
            )
        ready, _, _ = select.select([proc.stdout], [], [], remaining)
        if ready:
            line = proc.stdout.readline()
            if line.strip():
                return int(line.strip())


@pytest.fixture(scope="module")
def server(tmp_path_factory: pytest.TempPathFactory) -> Iterator[str]:
    data_dir = tmp_path_factory.mktemp("range_server_data")
    (data_dir / "small.fcb").symlink_to(SMALL)
    if DELFT.exists():
        (data_dir / "delft.fcb").symlink_to(DELFT)
    proc = subprocess.Popen(
        [sys.executable, str(RANGE_SERVER), str(data_dir)],
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        text=True,
    )
    try:
        port = _wait_for_port(proc)
        yield f"http://127.0.0.1:{port}"
    finally:
        proc.terminate()
        try:
            proc.wait(timeout=5)
        except subprocess.TimeoutExpired:
            proc.kill()
            proc.wait(timeout=5)


# ------------------------------------------------------------- the brief ---


def test_total_size_matches_the_local_file(server: str) -> None:
    r = HttpRangeReader(f"{server}/small.fcb")
    assert r.total_size() == SMALL.stat().st_size


def test_full_scan_over_http_matches_the_local_reader_exactly(
    server: str,
) -> None:
    local = FcbReader.open_file(SMALL)
    remote = FcbReader.open(HttpRangeReader(f"{server}/small.fcb"))

    local_features = [
        to_cityjson_feature(f, local.header) for f in local.select_all()
    ]
    remote_features = [
        to_cityjson_feature(f, remote.header) for f in remote.select_all()
    ]

    assert len(remote_features) == 3
    assert remote_features == local_features


def test_server_ignoring_range_raises_instead_of_misslicing(
    server: str,
) -> None:
    # The dangerous case the brief calls out by name: a server that
    # answers 200 with the whole body must be rejected, not silently
    # sliced as if it had honoured the Range header.
    r = HttpRangeReader(f"{server}/small.fcb?ignore_range=1")
    with pytest.raises(FcbError) as exc_info:
        r.total_size()
    assert exc_info.value.code is ErrorCode.IO_ERROR


# --------------------------------------------------- beyond the brief ---


def test_reads_match_the_local_file_byte_for_byte(server: str) -> None:
    local = FileRangeReader(SMALL)
    remote = HttpRangeReader(f"{server}/small.fcb")

    assert remote.total_size() == local.total_size()
    assert remote.read(0, 64) == local.read(0, 64)
    assert remote.read(1000, 256) == local.read(1000, 256)

    n = local.total_size()
    assert remote.read(n - 16, 16) == local.read(n - 16, 16)


def test_zero_length_read_never_contacts_the_server(server: str) -> None:
    r = HttpRangeReader(f"{server}/small.fcb")
    r.total_size()
    before = r.request_count
    assert r.read(100, 0) == b""
    assert r.request_count == before


def test_read_past_end_raises_out_of_bounds(server: str) -> None:
    # Python's FileRangeReader raises past EOF (a deliberate divergence
    # from the C++ reference, documented on FileRangeReader itself);
    # HttpRangeReader matches that Python-side convention, not C++'s.
    r = HttpRangeReader(f"{server}/small.fcb")
    n = r.total_size()
    with pytest.raises(FcbError) as exc_info:
        r.read(n + 1, 4)
    assert exc_info.value.code is ErrorCode.INDEX_OUT_OF_BOUNDS


def test_a_range_crossing_eof_returns_exactly_the_bytes_that_exist(
    server: str,
) -> None:
    r = HttpRangeReader(f"{server}/small.fcb")
    n = r.total_size()
    assert len(r.read(n - 10, 100)) == 10


def test_malformed_content_range_is_rejected(server: str) -> None:
    r = HttpRangeReader(f"{server}/small.fcb?bad_range=1")
    with pytest.raises(FcbError) as exc_info:
        r.total_size()
    assert exc_info.value.code is ErrorCode.IO_ERROR


def test_a_range_answering_the_wrong_offset_is_rejected(server: str) -> None:
    r = HttpRangeReader(f"{server}/small.fcb?wrong_offset=1")
    # total_size()'s own 1-byte probe (bytes=0-0) is too small for
    # wrong_offset's "start + 8 <= end" trigger, so it still succeeds;
    # the shift shows up on the next, larger read.
    r.total_size()
    with pytest.raises(FcbError) as exc_info:
        r.read(0, 64)
    assert exc_info.value.code is ErrorCode.IO_ERROR


def test_a_range_answering_an_over_long_end_is_rejected(server: str) -> None:
    # A server that matches the requested START but answers with an
    # END far past it (here: the rest of the whole file) must be
    # rejected, not trusted: `want = end - start + 1` is derived
    # straight from the server's Content-Range, so an unvalidated end
    # turns directly into an unbounded read for what should have been
    # total_size()'s tiny 1-byte probe (`Range: bytes=0-0`).
    r = HttpRangeReader(f"{server}/small.fcb?long_end=1")
    with pytest.raises(FcbError) as exc_info:
        r.total_size()
    assert exc_info.value.code is ErrorCode.IO_ERROR
    # Direct evidence this is the bounded rejection and not the
    # unbounded-read bug: nothing was ever buffered into
    # bytes_fetched. Pre-fix, this assertion also fails -- the whole
    # file ends up counted here instead.
    assert r.bytes_fetched == 0


def test_a_stalled_connection_mid_body_raises_io_error(server: str) -> None:
    # A server that sends 206 headers and then hangs before writing
    # any body must not let a raw stdlib exception (e.g. TimeoutError)
    # escape -- every network failure must surface as
    # FcbError(IO_ERROR), per the RangeReader contract
    # (range_reader.py:25). A short client-side timeout keeps this
    # test itself bounded; the server-side request thread left
    # blocked is torn down by the `server` fixture's process-kill
    # teardown, not by anything here.
    r = HttpRangeReader(f"{server}/small.fcb?stall_body=1", timeout=0.5)
    with pytest.raises(FcbError) as exc_info:
        r.total_size()
    assert exc_info.value.code is ErrorCode.IO_ERROR


@pytest.mark.skipif(not DELFT.exists(), reason="delft.fcb fixture missing")
def test_attribute_query_over_http_matches_the_local_reader(
    server: str,
) -> None:
    query = [AttrCondition("b3_bouwlagen", Operator.GE, KeyValue.from_u64(1))]
    local = FcbReader.open_file(DELFT)
    remote = FcbReader.open(HttpRangeReader(f"{server}/delft.fcb"))

    local_offsets = sorted(h.offset for h in local.select_attr(query))
    remote_offsets = sorted(h.offset for h in remote.select_attr(query))

    assert len(remote_offsets) > 0
    assert remote_offsets == local_offsets


@pytest.mark.skipif(not DELFT.exists(), reason="delft.fcb fixture missing")
def test_http_reader_issues_ranged_requests_not_the_whole_file(
    server: str,
) -> None:
    # The entire point of a RangeReader: a query over a 7+ MB remote
    # file must fetch a small fraction of it, via more than one ranged
    # request -- never one big non-range GET of the whole body.
    query = [AttrCondition("b3_bouwlagen", Operator.GE, KeyValue.from_u64(1))]
    reader = HttpRangeReader(f"{server}/delft.fcb")
    total = reader.total_size()

    fcb = FcbReader.open(reader)
    hits = list(fcb.select_attr(query))

    assert hits
    assert reader.request_count > 1
    assert 0 < reader.bytes_fetched < total


# ---------------------------------------------------- live remote (opt-in) ---


@pytest.mark.skipif(
    not REMOTE_URL,
    reason="set FCB_REMOTE_HTTP_URL to run (see `just test-remote`)",
)
def test_remote_3dbag_opens_and_queries_over_http() -> None:
    """The published 3DBAG file, read over real HTTP range requests.

    Proves three things at once: the header passes verification (the file
    is in the post-alignment-fix format), the whole 68 GB is never
    downloaded (a bounded number of ranged requests), and the spatial
    index traverses to the same feature set the other three readers see.
    """
    reader = HttpRangeReader(REMOTE_URL)
    fcb = FcbReader.open(reader)
    header = fcb.header

    assert header.info.features_count == REMOTE_FEATURES_COUNT

    # Opening a 68 GB file must cost a handful of small ranged reads, not a
    # download. If this ever approaches total_size the range logic regressed.
    assert 0 < reader.bytes_fetched < 5_000_000

    hits = search_rtree(
        fcb.range_reader,
        header.layout.rtree_begin,
        header.info.features_count,
        header.info.index_node_size,
        REMOTE_BBOX,
    )
    # Exact, and identical to what the Rust, C++ and TypeScript readers
    # return for the same box -- see the constants' note above.
    assert len(hits) == REMOTE_BBOX_COUNT

    # The bbox scan reads more, but still a tiny fraction of the file.
    assert reader.bytes_fetched < total_size_guard(reader)


def total_size_guard(reader: HttpRangeReader) -> int:
    """A generous ceiling: 1% of the file. The Amsterdam box touches far
    less, but pinning an exact byte count would be brittle against future
    re-serializations, whereas 'never more than 1% of 68 GB' is a stable
    statement of the property under test."""
    return reader.total_size() // 100
